[Guided onboarding] Implement quit guide functionality

alisonelizabeth opened the pull request:
This PR implements the "Exit setup guide" functionality in the guide dropdown panel. Once the user confirms to quit a guide, an API request is made to delete the SO that is storing the guide's state.

Fixes https://github.com/elastic/kibana/issues/139804

### How to test
1. Enable guided onboarding via your `kibana.dev.yml` file: `guidedOnboarding.ui: true`
2. Start up Kibana with the example plugins enabled: `yarn start --run-examples`
3. Navigate to `/app/guidedOnboardingExample`
4. Click the "Start search guide" button to trigger the guide
5. Click the "Quit setup guide" link at the button of the dropdown panel and verify modal displays.
6. Confirm clicking "Cancel" or the "x" icon closes the modal and maintains the guide state.
7. Confirm clicking "Quit guide" exits the guide. The Setup guide button should be disabled (behavior will change once https://github.com/elastic/kibana/issues/141129 is implemented). Also, if you navigate back to `/app/guidedOnboardingExample` the Search guide should say "Continue".


### Screenshots

<img width="555" alt="Screen Shot 2022-10-10 at 9 31 13 AM" src="https://user-images.githubusercontent.com/5226211/194878170-2a48fba0-ead8-470c-83b5-20aa11af2154.png">


// Loading state while API request is in progress
<img width="587" alt="Screen Shot 2022-10-05 at 9 04 00 AM" src="https://user-images.githubusercontent.com/5226211/194073565-beef1cfc-ada3-49bd-ae4c-0b9b6eb6e799.png">

gchaps commented:
@kellyemurphy can you please review the copy in this PR?

kellyemurphy commented on src/plugins/guided_onboarding/public/components/quit_guide_modal.tsx:
```diff
@@ -0,0 +1,98 @@
+      setIsDeleting(false);
+      notifications.toasts.addError(error, {
+        title: i18n.translate('guidedOnboarding.quitGuideModal.errorToastTitle', {
+          defaultMessage: 'There was an error quitting the guide. Please try again.',
```
Do we want them to try again right away or should they wait a moment?

alisonelizabeth replied:
Hmm... good question. I suppose it depends on the actual error. Do we have guidelines around error messages when an API request fails?

kellyemurphy replied:
The error message guidance that we've provided the platform (cloud) team in the past is that we want to A. tell them what happened, B. tell them what they can do about it, and C. only use please when we are asking for their patience. This message meets those guidelines, but it's just a matter of timing.

kellyemurphy commented on src/plugins/guided_onboarding/public/components/quit_guide_modal.tsx:
```diff
@@ -0,0 +1,98 @@
+        <EuiText>
+          <p>
+            {i18n.translate('guidedOnboarding.quitGuideModal.modalDescription', {
+              defaultMessage: 'You can restart anytime, just click Setup guide on the homepage.',
```
```suggestion
              defaultMessage: 'You can restart anytime, just open the Setup guide from the Help menu.',
```

Recommending this tweak due to [the button location in this issue](https://github.com/elastic/kibana/issues/139816).

kellyemurphy commented on src/plugins/guided_onboarding/public/components/quit_guide_modal.tsx:
```diff
@@ -0,0 +1,98 @@
+        <EuiTitle size="m">
+          <h2>
+            {i18n.translate('guidedOnboarding.quitGuideModal.modalTitle', {
+              defaultMessage: 'Quit this setup guide and discard progress?',
```
```suggestion
              defaultMessage: 'Quit this guide and discard progress?',
```
I think we can tweak this a bit more, since they already know they're in a setup guide.

elasticmachine commented:
Pinging @elastic/platform-onboarding (Team:Journey/Onboarding)

cindychangy reviewed (approved):
Thanks @alisonelizabeth, LGTM!

yuliacech reviewed (commented):
Thanks a lot for implementing this functionality, @alisonelizabeth! 
The UI changes LGTM but I'm not sure if we should delete the progress of the guide though. I thought when the use quits the guide, the state of the guided onboarding is set to something like "inactive" so that we know that the button is not shown in the header anymore. But if the user comes back to the landing page, they can see their previous progress on all guides. And they could continue a guide from the step where they left. 
If we delete the state, I'm not sure how we would handle when the user comes back to the landing page and tries to restart a guide, wdyt?

alisonelizabeth commented:
> The UI changes LGTM but I'm not sure if we should delete the progress of the guide though. I thought when the use quits the guide, the state of the guided onboarding is set to something like "inactive" so that we know that the button is not shown in the header anymore. But if the user comes back to the landing page, they can see their previous progress on all guides. And they could continue a guide from the step where they left.
> If we delete the state, I'm not sure how we would handle when the user comes back to the landing page and tries to restart a guide, wdyt?

That's a great question. @osmanis @cindychangy @bojanasan can you weigh in here? Based on the copy ("Quit this guide and discard progress?"), I thought the state would be discarded. However, I can't seem to find that in the design spec, so I might have made an assumption there 😅.

cindychangy commented:
Good catch, @yuliacech is correct that when a user quits the guide - it should save their state. We have a [page on the prototype](https://cindychangy.github.io/platform-onboarding/8.6/guided-setup/in-progress/) to show this situation. If a user re-visits the landing page they should see their progress of all the guides, and the buttons will show different text based on that ("Start", "Continue"...). @alisonelizabeth: I agree given this, the copy should be tweaked as "discard state" might be confusing if we are saving their state. @kellyemurphy do you have a suggestion on how to best tweak the title message: "Quit this guide and discard progress?".

alisonelizabeth commented:
Thanks @cindychangy! For now, I've changed the modal title to simply: "Quit this guide?". @kellyemurphy let me know if you'd like me to make any further changes.

@yuliacech would you mind taking another look when you have time? I updated the functionality via https://github.com/elastic/kibana/pull/142542/commits/dc79bcea9ea7c93c5ba09d41b3774ace514cbfaa.

yuliacech reviewed (approved):
Thanks a lot, @alisonelizabeth! Latest changes LGTM 👍

alisonelizabeth commented:
I'm going to go ahead and merge this PR. I can follow-up with a separate PR if there are any additional copy changes.
